src/main/java/com/google/cloud/tools/managedcloudsdk/internal/download/StandardDownloader.java

```diff
@@ -0,0 +1,104 @@
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.NotDirectoryException;
+import java.nio.file.Path;
+
+/** Standard implementation of {@link Downloader}. */
+public final class StandardDownloader implements Downloader {
+
+  static final int BUFFER_SIZE = 10 * 1024;
+  private final URL address;
+  private final Path destinationFile;
+  private final String userAgentString;
+  private final DownloadProgressListener downloadProgressListener;
+
+  /** Use {@link DownloaderFactory} to instantiate. */
+  StandardDownloader(
+      URL source,
+      Path destinationFile,
+      String userAgentString,
+      DownloadProgressListener downloadProgressListener) {
+    this.address = source;
+    this.destinationFile = destinationFile;
+    this.userAgentString = userAgentString;
+    this.downloadProgressListener = downloadProgressListener;
+  }
+
+  @Override
+  public Path call() throws IOException, InterruptedException {
+    Path destinationDir = destinationFile.getParent();
+    if (Files.exists(destinationDir) && !Files.isDirectory(destinationDir)) {
+      throw new NotDirectoryException(
+          "Cannot download to " + destinationDir + " because it is not a directory");
+    }
+    if (Files.exists(destinationFile)) {
+      throw new FileAlreadyExistsException(
+          "Cannot write to " + destinationFile + " because it already exists");
+    }
+    if (!Files.exists(destinationDir)) {
+      Files.createDirectories(destinationDir);
+    }
+
+    URLConnection conn;
+    conn = address.openConnection();
+    conn.setRequestProperty("User-Agent", userAgentString);
+
+    try (BufferedOutputStream out =
+        new BufferedOutputStream(Files.newOutputStream(destinationFile))) {
+      try (InputStream in = conn.getInputStream()) {
+
+        long contentLength = conn.getContentLengthLong();
+        // Progress is updated every 1%
+        long updateThreshold = contentLength / 100;
+
+        byte[] buffer = new byte[BUFFER_SIZE];
+        long lastUpdated = 0;
+        long totalBytesRead = 0;
+
+        int bytesRead;
+        while ((bytesRead = in.read(buffer)) != -1) {
+          if (Thread.currentThread().isInterrupted()) {
+            throw new InterruptedException("Downloader was interrupted.");
+          }
+          totalBytesRead += bytesRead;
+          if (totalBytesRead - lastUpdated > updateThreshold) {
+            if (downloadProgressListener != null) {
+              downloadProgressListener.updateProgress(
+                  totalBytesRead - lastUpdated, totalBytesRead, contentLength);
+            }
+            lastUpdated = totalBytesRead;
+          }
+          out.write(buffer, 0, bytesRead);
+        }
+        if (downloadProgressListener != null && (lastUpdated != totalBytesRead)) {
```

Comment:
This can be combined to the one above by changing the expression on line 87 to like

`totalBytesRead == contentLength || totalBytesRead - lastUpdated > updateThreshold`

Reply:
yup, much cleaner.